epochStart/metachain/sovereignRewards.go

```diff
@@ -0,0 +1,201 @@
+
+	sr.fillBaseRewardsPerBlockPerNodeFunc = sr.fillBaseRewardsPerBlockPerNode
+	sr.flagDelegationSystemSCEnabled.SetValue(true)
+	return sr, nil
+}
+
+// CreateRewardsMiniBlocks creates the rewards miniblocks according to economics data and validator info.
+// This method applies the rewards according to the economics version 2 proposal, which takes into consideration
+// stake top-up values per node. For each created reward tx, it will add directly to the reward txs pool
+func (rc *sovereignRewards) CreateRewardsMiniBlocks(
+	metaBlock data.MetaHeaderHandler,
+	validatorsInfo state.ShardValidatorsInfoMapHandler,
+	computedEconomics *block.Economics,
+) (block.MiniBlockSlice, error) {
+	if check.IfNil(metaBlock) {
+		return nil, epochStart.ErrNilHeaderHandler
+	}
+	if computedEconomics == nil {
+		return nil, epochStart.ErrNilEconomicsData
+	}
+
+	rc.mutRewardsData.Lock()
+	defer rc.mutRewardsData.Unlock()
+
+	log.Debug("rewardsCreatorV2.CreateRewardsMiniBlocks",
+		"totalToDistribute", computedEconomics.TotalToDistribute,
+		"rewardsForProtocolSustainability", computedEconomics.RewardsForProtocolSustainability,
+		"rewardsPerBlock", computedEconomics.RewardsPerBlock,
+		"devFeesInEpoch", metaBlock.GetDevFeesInEpoch(),
+		"rewardsForBlocks no fees", rc.economicsDataProvider.RewardsToBeDistributedForBlocks(),
+		"numberOfBlocks", rc.economicsDataProvider.NumberOfBlocks(),
+		"numberOfBlocksPerShard", rc.economicsDataProvider.NumberOfBlocksPerShard(),
+	)
+
+	miniBlocks := rc.initializeRewardsMiniBlocks()
+	rc.clean()
+
+	protRwdTx, protRwdShardId, err := rc.createProtocolSustainabilityRewardTransaction(metaBlock, computedEconomics)
+	if err != nil {
+		return nil, err
+	}
+
+	nodesRewardInfo, dustFromRewardsPerNode := rc.computeRewardsPerNode(validatorsInfo)
+	log.Debug("arithmetic difference from dust rewards per node", "value", dustFromRewardsPerNode)
+
+	dust, err := rc.addValidatorRewardsToMiniBlocks(metaBlock, miniBlocks, nodesRewardInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	dust.Add(dust, dustFromRewardsPerNode)
+	log.Debug("accumulated dust for protocol sustainability", "value", dust)
+
+	rc.adjustProtocolSustainabilityRewards(protRwdTx, dust)
+	err = rc.addProtocolRewardToMiniBlocks(protRwdTx, miniBlocks, protRwdShardId)
+	if err != nil {
+		return nil, err
+	}
+
+	return rc.finalizeMiniBlocks(miniBlocks), nil
+}
+
+func (rc *sovereignRewards) addValidatorRewardsToMiniBlocks(
+	metaBlock data.HeaderHandler,
+	miniBlocks block.MiniBlockSlice,
+	nodesRewardInfo map[uint32][]*nodeRewardsData,
+) (*big.Int, error) {
+	rwdAddrValidatorInfo, accumulatedDust := rc.computeValidatorInfoPerRewardAddress(nodesRewardInfo)
+
+	for _, rwdInfo := range rwdAddrValidatorInfo {
+		rwdTx, rwdTxHash, err := rc.createRewardFromRwdInfo(rwdInfo, metaBlock)
+		if err != nil {
+			return nil, err
+		}
+		if rwdTx.Value.Cmp(zero) <= 0 {
+			continue
+		}
+
+		rc.accumulatedRewards.Add(rc.accumulatedRewards, rwdTx.Value)
+
+		if rwdTx.Value.Cmp(zero) < 0 {
```

Comment:
This `if` will never be `true`. You can log the error at line 108 if you want.

Reply:
Right, fixed it